### PR TITLE
Components: Fix DateTimePicker timezone handling for non-string values

### LINKED ISSUE #73981: Calendar in "publish" popover no longer shows dots for published posts
### Description

The dots in the calendar showing other post dates are missing, likely due to changes in #73444. (There are also some fixes in progress at #73887, which may or may not be related.)

### Step-by-step reproduction instructions

1. Have one or two posts published in the current month.
2. Go to Posts → Add New
3. In the Post sidebar, click on "Publish: Immediately" (or the date if already set)
4. The calendar date picker opens

### Expected: Days with existing published posts should show small dots beneath the date number

Dot in calendar

### Actual: No dots appear on any dates

No dot in calendar

### Screenshots, screen recording, code snippet

_No response_

### Environment info

Latest trunk

### Please confirm that you have searched existing issues in the repo.

- [x] Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

- [x] Yes

### Please confirm which theme type you used for testing.

- [x] Block
- [ ] Classic
- [ ] Hybrid (e.g. classic with theme.json)
- [ ] Not sure

### PR DESCRIPTION
**Pull request by aduth**
## What?

Fixes an issue with `DateTimePicker` after #73444 where `Date` or `number` or `string` (with timezone) values passed as `currentDate` would be treated inconsistently from timezoneless strings, potentially resulting in mismatches between displayed and changed values similar to what was seen previously in #67480.

Related to: #67480
Follow-up to #73444
Fixes #73981

## Why?

The intent of the changes with #73444 were to normalize internal handling of dates within the component to use UTC time, but the changes only applied to values passed as timezoneless strings. In current Gutenberg usage, this is primarily how the component is used, since this is how WordPress represents dates.

Since the component internally assumes UTC times, this behavior should also extend to values passed in other formats (date objects, timestamps).

## How?

Updates `inputToDate` to consistently coerce all input values to UTC time.

## Testing Instructions

Verify included tests pass:

- `npm run test:unit packages/components/src/date-time/date-time/test/index.tsx`
- `npm run test:unit packages/components/src/date-time/test/utils.test.ts`

I was able to identify one Gutenberg usage where this bug previously existed, in [the initialization of the "Date" block](https://github.com/WordPress/gutenberg/blob/90c242a3020dc2ed8865cff38e5a33d7aa318fbd/packages/block-library/src/post-date/edit.js#L77).

To verify:

1. Set your local system timezone to a different date from the site's timezone (see instructions in #73444)
2. Go to "Posts" > "Add Post"
3. Insert a "Date" block
4. In the block toolbar, click "Change Date"
5. Verify that the initial date is accurately displayed as the current time on the site

## Screenshots or screencast \

These screenshots show a before/after of the Date block instructions described above, where my local time (1:52am UTC+10.5 the next day) is incorrectly shown, compared to the site's actual current time (3:22pm UTC).

Before|After
---|---
Screenshot 2025-12-11 at 1 52 31 AM|Screenshot 2025-12-11 at 1 52 54 AM

### CODE REVIEW

File: packages/components/src/date-time/utils.ts [hunk 1]
```diff
@@ -27,9 +25,13 @@ export function inputToDate( input: Date | string | number ): Date {
 		//
 		// See: https://tc39.es/ecma262/#sec-date-time-string-format
 		const hasTimezone = /Z|[+-]\d{2}(:?\d{2})?$/.test( input );
-		return hasTimezone ? new Date( input ) : new UTCDateMini( input + 'Z' );
+		return new UTCDateMini( hasTimezone ? new Date( input ) : input + 'Z' );
```
Comment (tyxla) [inline packages/components/src/date-time/utils.ts]: Can this break any existing use cases? Do you feel like we're well covered by tests?

File: packages/components/src/date-time/utils.ts [hunk 2]
```diff
@@ -27,9 +25,13 @@ export function inputToDate( input: Date | string | number ): Date {
...
 	}
-	return toDate( input );
+
+	// Date objects and number timestamps represent specific UTC moments.
+	// Convert to milliseconds since epoch for consistent UTC handling.
+	const time = input instanceof Date ? input.getTime() : input;
+	return new UTCDateMini( time );
```
Comment (tyxla) [inline packages/components/src/date-time/utils.ts:46]: Same question as above

File: packages/components/src/date-time/utils.ts (hunk 1 above)
Comment (aduth) [reply packages/components/src/date-time/utils.ts]: These changes should un-break some of the assumptions introduced by changes in #73444 for values other than timezoneless strings, since the changes there start introducing assumptions around UTC times in how dates are displayed. Although curiously none of these changes have affected how those date changes are serialized, so I'd be curious to test if this has always been somewhat broken for these types of date values.

As far as tests are concerned, there's a lot of different factors like local vs. server timezones, date object shapes, daylight savings time, etc, but I'd like to think (especially after the changes here) we have exhaustive branch coverage. For example, we previously had no unit test coverage for `inputToDate` covering timestamps.

Comment (tyxla) [reply packages/components/src/date-time/utils.ts]: Maybe we can run some AI prompts to see if any additional coverage ideas come up?

Comment (aduth) [reply packages/components/src/date-time/utils.ts]: Here's what AI had to say:

>### Coverage Assessment
>
>| Test Category | Coverage | Notes |
>|---------------|----------|-------|
>| String inputs (timezoneless) | ✅ Excellent | Multiple timezone scenarios across unit + integration tests |
>| String inputs (with timezone) | ✅ Good | Z, +offset, -offset all covered |
>| Date object inputs | ✅ Good | Unit test + DateTimePicker integration test |
>| Timestamp inputs | ✅ Good | Unit test + DateTimePicker integration test |
>| Timezone edge cases | ✅ Excellent | DST transitions, timezone mismatches tested |
>| Component usage in DatePicker | ⚠️ Adequate | Tests use strings; no explicit Date/timestamp testing |
>| Component usage in TimePicker | ⚠️ Adequate | Tests use strings; no explicit Date/timestamp testing |
>
>### Potential Gaps
>
>1. **TimePicker with non-string inputs**: While `DateTimePicker` tests Date objects and timestamps, the standalone `TimePicker` tests only use string inputs. However, since `TimePicker` uses the same `inputToDate` function and it's tested at the unit level, this is likely acceptable.
>
>2. **DatePicker with non-string inputs**: Same as above — tests only use strings. Again, covered by unit tests.
>
>3. **Invalid input handling**: No explicit tests for malformed inputs (e.g., invalid date strings). The function doesn't have explicit error handling, so this may be intentional.
>
>### Verdict
>
>**The test coverage is quite exhaustive for protecting against regressions**, particularly because:
>
>1. The PR commits (`718c5334126` and `13dcb543d54`) specifically added comprehensive timezone handling tests
>2. The unit tests mock multiple timezones to catch day-shift bugs
>3. The `DateTimePicker` integration tests cover the most critical real-world scenarios with timezone mismatches and DST transitions
>4. All three input types (strings, Date objects, timestamps) are tested
>
>The tests are well-designed to catch timezone-related regressions, which appears to be the core concern of PR 73887's fix for "DateTimePicker timezone handling for non-string values."

**Review by tyxla — APPROVED** [review]
Thanks @aduth, let's get this in 👍

**Comment by aduth** [comment]
After some more testing and reports, I discovered that there were a lot more issues with the implementation that needed to be solved. This required more extensive changes than initially expected, though I tried to keep the surface area as small as possible. In some cases, it reverts to some logic that had existed prior to #73444 (e.g. using `dateI18n` instead of `gmdateI18n` for date display).

A lot of the problems stem back to the fact that we're juggling three different timezones representations: the browser's local timezone (what's used with `new Date()`), UTC (useful as a normalization format), and the site's configured timezone.

One of the challenges is in how we interpret the different types of values. WordPress primarily uses timezoneless strings which map to the site's configured timezones, but the component also supports receiving `Date` values which are likely to be initialized using the browser's local timezone.

Here's how I think the values should be interpreted:

Input type | Example | Interpreted as | Displayed as
---|---|---|---
Timezoneless string | `"2025-11-15T20:00:00"` | Represents 20:00 in site timezone | Shows Nov 15, 20:00 regardless of browser timezone
Date object | `new Date(Date.UTC(2025, 10, 16, 1, 0, 0))` | UTC instant (Nov 16 01:00 UTC) | Converted to site timezone for display (Nov 15 20:00 if site is UTC-5)
Timestamp | `1731718800000` | UTC instant | Converted to site timezone for display
Default | `undefined` | Defaults to current time | Interpreted as current time in site timezone

Recapping all of what's been fixed in the recent changes:

- Date/time drift on round-trips: Values could shift when passed through `onChange` and back to the component. For example, components which pass `Date` values would likely be converting the received `onChange` string value back to a `Date` object.
- "Today" highlighting: This was based on browser timezone instead of site timezone
- Event indicators (#73981): These were associated with calendar days based on browser timezone instead of site timezone
- `undefined` input handling: Similarly used browser's current time instead of site's current time, which could cause the wrong day to be highlighted as "Today"
- `@wordpress/date`'s handling of `0` numeric timezone: These functions were checking truthiness to indicate the presence of a timezone argument (wrong behavior for `0` values, which are falsey but should be interpreted as present)

A screenshot showing the before/after effect on "Today" handling and scheduled events:

Before|After
---|---
Screenshot 2025-12-18 at 12 28 45 AM|Screenshot 2025-12-18 at 12 29 10 AM

I've expanded test coverage to include:

- Handling `undefined` initial dates
- Transforming values on `onChange` to maintain the expected original type
- Asserting for indicators on "Today" and "There are # events"

Regarding "Today", this was challenging to test for because there were only visual style indicators for the effect, and no semantic representation. There wasn't even a deterministic class name to look for. As an accessibility improvement, I chose to change the label to include "Today" as part of the date label if it is the current date, similar to what we do with "Selected" label suffix for the selected date.

**Comment by aduth** [comment]
There were still yet more issues after the last round of changes, particularly around assigning times. This was something we didn't have great test coverage for, which is vastly expanded in 2f1cb7a86f0214f36bc84c80a6e8f4f5a2edf3fc for both the `TimePicker` component itself, as well as composition within the `DateTimePicker` component. 

Broad improvements to test coverage:

- Matrix nesting for combinations of both site before/same/after local timezone and different input types
- Mirror behavior coverage between `DateTimePicker` and `TimePicker`
   - Include time behavior testing in `DateTimePicker` tests

There were two issues with how we were managing time:
- The component was displaying the times correctly according to the site configured timezone, but the change handler would output times based on the browser's local timezone, which could lead to time drift if the site and local timezones differed.
- The component internally used `date-fn`'s `set` helper, but this operates on local date values rather than the internally-normalized UTC dates.

#73444 seems to have highlighted some fundamental issues with how the component would deal with values other than strings. It seemed to largely operate on local time values, but timezoneless strings are meant to represent time in the site's configured timezone. The changes that have been introduced seek to enforce this. In retrospect, it probably would have been a wiser choice to align the `onChange` type to the input type, because `new Date( timezonelessString )` produces the wrong result when the configured timezone is not the same as the local browser timezone.

**Review by jasperck — APPROVED** [review]
Thanks a bunch! I tested from [here](https://github.com/Automattic/dayone-web-client/pull/7), and everything’s looking good now 🎉

**Comment by aduth** [comment]
In some final testing, I discovered a couple other subtle bugs. These were not quite as severe as others we've seen, but emblematic of the same root causes:

1. <kbd>Home</kbd> and <kbd>End</kbd> are meant to select the start and end days of the week, but these were incorrectly selecting Saturday/Sunday in some cases where it should have been selecting Monday/Sunday. 
   - I discovered these in an audit of usage of `startOf*` and `endOf*` functions from `date-fns`, as the way these operate as assuming local timezone is incompatible with how we internally treat dates as UTC-normalized. Sure enough, the remaining usage of the functions resulted in unexpected behaviors.
   - Fixed with regression coverage in b87eeb951335c203e3a182e14b4b726a8f7c7b40
2. Event dots would sometimes have off-by-one errors on the date that they were displayed, e.g. a post scheduled for 2025-12-20 13:00 UTC with site configured in UTC time would show a dot on the 19th if the system time was configured in Adelaide (UTC+09:30). 
   -  As elaborated in the extended commit description of 2533afa87ce6cc05221b5ca3441ba38cf887e461, this is another case where [constructing `Date` objects using timezoneless strings](https://github.com/WordPress/gutenberg/blob/14b4649d18aefb78596c2aafc014d870ae4790c5/packages/editor/src/components/post-schedule/index.js#L79) is fraught, because `Date` objects reflect the local timezone whereas timezoneless strings reflect the site's configured timezone, and these are not necessarily the same.
   - Fixed with regression coverage in 2533afa87ce6cc05221b5ca3441ba38cf887e461

**Comment by aduth** [comment]
In testing across this branch and what's in the stable release of WordPress 6.9, one other thing I noticed that this fixes is that the "Today" highlight in 6.9 follows the current browser time rather than the site time. The screenshots below show a scenario where my local time (02:38 2025-12-20UTC+9:30) is a day ahead of the site time (15:59 2025-12-19UTC). My expectation is that this highlight should follow what "Today" is in the site's time (i.e. the 19th), same as how we default to the "current" time in scheduling.

I couldn't find an open issue related to this, but raising for awareness.

State|Before (WordPress 6.9)|After
---|---|---
Initial page load|image|image
Selecting 18th|image|image
